In [114]:
import numpy as np

def coin_toss(p):     
    result = np.random.binomial(1,p) 
    if result == 0:
        experiment = 'T'
    else :
        experiment = 'H'
    return experiment


def position_experiments(): 
    p=0.5
    n=100       # number of tosses in one experiment
    full_experiments = []

    for i in range(0, n):    
        full_experiments.append(coin_toss(p))
        i+=1
    
    #print(full_experiments)
    
    htt, hth = 0,0
    
    for i in range(0, len(full_experiments)-3):
        if full_experiments[i] == 'H' and full_experiments[i+1] == 'T' and full_experiments[i+2] == 'T' and htt == 0:
            htt = i+1
        elif full_experiments[i] == 'H' and full_experiments[i+1] == 'T' and full_experiments[i+2] == 'H' and hth == 0:
            hth = i+1
        
        if htt != 0 and hth != 0 :
            break
        i+=1
    
    #print('HTT' + str(htt))
    #print('HTH' + str(hth))
    return [htt, hth]


def Average(lst): 
    return sum(lst) / len(lst)

In [117]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

print('The list of numbers of tosses until HTT :')
print(HTT_position)
print()
print('The list of numbers of tosses until HTH :')
print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print()
print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The list of numbers of tosses until HTT :
[6, 14, 1, 10, 5, 9, 4, 3, 10, 14, 4, 10, 4, 2, 2, 5, 14, 1, 4, 7, 4, 9, 15, 4, 2, 5, 9, 5, 11, 16, 20, 2, 3, 4, 2, 5, 1, 17, 6, 6, 10, 4, 2, 2, 6, 16, 12, 12, 1, 7, 7, 2, 6, 3, 4, 1, 4, 3, 8, 12, 10, 7, 11, 4, 5, 4, 1, 11, 1, 4, 1, 6, 1, 5, 10, 6, 5, 4, 2, 2, 1, 6, 2, 3, 4, 15, 1, 6, 12, 5, 9, 10, 2, 6, 3, 1, 10, 2, 2, 3, 2, 1, 7, 6, 7, 6, 4, 2, 3, 3, 2, 7, 5, 5, 5, 5, 4, 10, 2, 5, 1, 7, 3, 11, 2, 1, 6, 5, 3, 9, 4, 10, 4, 1, 5, 3, 2, 1, 5, 1, 1, 8, 5, 7, 9, 6, 16, 7, 21, 9, 1, 5, 4, 14, 7, 1, 1, 11, 2, 3, 3, 6, 3, 12, 4, 4, 2, 2, 9, 7, 11, 2, 15, 8, 2, 4, 4, 2, 2, 10, 3, 4, 2, 4, 3, 8, 11, 9, 5, 5, 17, 5, 4, 19, 2, 5, 11, 8, 12, 14, 9, 2, 10, 2, 5, 14, 3, 5, 3, 7, 5, 1, 3, 10, 18, 5, 2, 2, 7, 17, 2, 4, 11, 2, 5, 3, 5, 3, 14, 6, 12, 1, 2, 3, 7, 2, 6, 1, 3, 4, 5, 2, 17, 3, 6, 2, 5, 3, 5, 10, 4, 1, 2, 7, 2, 7, 10, 5, 1, 1, 9, 12, 7, 2, 2, 6, 5, 14, 4, 9, 4, 2, 6, 18, 16, 7, 3, 2, 4, 3, 2, 3, 11, 2, 10, 8, 2, 3, 6, 4, 3, 3, 7, 1, 15, 7, 1, 3, 9, 3

In [110]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

# print('The list of numbers of tosses until HTT :')
# print(HTT_position)
# print('The list of numbers of tosses until HTH :')
# print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The average number of tosses until HTT : 5.946
The average number of tosses until HTH : 8.118


In [111]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

# print('The list of numbers of tosses until HTT :')
# print(HTT_position)
# print('The list of numbers of tosses until HTH :')
# print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The average number of tosses until HTT : 6.13
The average number of tosses until HTH : 8.195


In [112]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

# print('The list of numbers of tosses until HTT :')
# print(HTT_position)
# print('The list of numbers of tosses until HTH :')
# print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The average number of tosses until HTT : 6.098
The average number of tosses until HTH : 7.869


In [113]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

# print('The list of numbers of tosses until HTT :')
# print(HTT_position)
# print('The list of numbers of tosses until HTH :')
# print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The average number of tosses until HTT : 6.3
The average number of tosses until HTH : 7.855


In [118]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

# print('The list of numbers of tosses until HTT :')
# print(HTT_position)
# print('The list of numbers of tosses until HTH :')
# print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The average number of tosses until HTT : 6.046
The average number of tosses until HTH : 8.035


In [119]:
HTT_position = []
HTH_position = []

e=1000          # number of experiments
for n in range(0, e): 
    list = position_experiments()
    HTT_position.append(list[0])
    HTH_position.append(list[1])

# print('The list of numbers of tosses until HTT :')
# print(HTT_position)
# print('The list of numbers of tosses until HTH :')
# print(HTH_position)

avg_htt = Average(HTT_position) 
avg_hth = Average(HTH_position) 

print('The average number of tosses until HTT : ' + str(avg_htt))
print('The average number of tosses until HTH : ' + str(avg_hth))

The average number of tosses until HTT : 5.98
The average number of tosses until HTH : 8.081


# Report

For this problem, I run 1000 experiments with 100 tosses each. Moreover, I also run this code 7 times. The result is as follows :

First run :
<br>
The average number of tosses until HTT : 5.886
<br>
The average number of tosses until HTH : 8.063

Second run :
<br>
The average number of tosses until HTT : 5.946
<br>
The average number of tosses until HTH : 8.118

Third run :
<br>
The average number of tosses until HTT : 6.13
<br>
The average number of tosses until HTH : 8.195

Fourth run :
<br>
The average number of tosses until HTT : 6.098
<br>
The average number of tosses until HTH : 7.869

Fith run :
<br>
The average number of tosses until HTT : 6.3
<br>
The average number of tosses until HTH : 7.855

Sixth run :
<br>
The average number of tosses until HTT : 6.046
<br>
The average number of tosses until HTH : 8.035

Seventh run :
<br>
The average number of tosses until HTT : 5.98
<br>
The average number of tosses until HTH : 8.081

## Conclusion
It can be seen that the average number of tosses until HTH is larger than the average number of tosses until HTT, especially when we have a large number of experiments (converge).